apps/snitch_core/lib/core/data/schema/product.ex

```diff
@@ -0,0 +1,44 @@
+defmodule Snitch.Data.Schema.Product do
+  @moduledoc """
+  Models a Product.
+  """
+
+  use Snitch.Data.Schema
+  alias Snitch.Data.Schema.Product.NameSlug
+  alias Snitch.Data.Schema.Variant
+
+  @type t :: %__MODULE__{}
+
+  schema "snitch_products" do
+    field(:name, :string, null: false, default: "")
+    field(:description, :string)
+    field(:available_on, :utc_datetime)
+    field(:deleted_at, :utc_datetime)
+    field(:discontinue_on, :utc_datetime)
+    field(:slug, :string)
+    field(:meta_description, :string)
+    field(:meta_keywords, :string)
+    field(:meta_title, :string)
+    field(:promotionable, :boolean)
+    timestamps()
+
+    has_many(:variants, Variant)
+  end
+
+  @required_fields ~w(name)a
+  @optional_fields ~w(description meta_description meta_keywords meta_title)a
+
+  def changeset(model, params \\ %{}) do
+    model
+    |> cast(params, @required_fields ++ @optional_fields)
+    |> validate_required(@required_fields)
+    |> NameSlug.maybe_generate_slug()
+    |> NameSlug.unique_constraint()
+  end
+end
+
+defmodule Snitch.Data.Schema.Product.NameSlug do
```

Comment:
@SagarKarwande, In my opinion, we can have many entities in the system which may need a slug and mostly slugs are derived from the name of the `model` only. So, I would recommend extracting this functionality such that it can be used across entities. As an effect, this module should become reusable. What do you think? @aviabird/snitch

Reply:
IMO there's not much scope for reuse as snitch_core runs on just IDs and UUIDs. Slugification is an artefact of a frontend and the controller so I don't want to keep slug fields in the DB. Existing ones can be removed.

Reply:
@oyeb Slug is like an identifiable surrogate key. It is unique and stored in the DB.